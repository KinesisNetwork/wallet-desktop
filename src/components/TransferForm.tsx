import * as copy from 'copy-to-clipboard'
import * as React from 'react'

import { TransferProps } from '@containers/TransferForm'
import { formAlert } from '@helpers/alert'
import { InputError, WalletLockError } from '@helpers/errors'
import { generateTransferTransaction } from '@services/transfer'
import { InputField } from './InputField'
import { Loader } from './Loader'
import { PayeeSelector } from './PayeeSelector'

export class TransferForm extends React.Component<TransferProps> {
  initTransfer = async (ev: React.MouseEvent<HTMLButtonElement>) => {
    ev.preventDefault()
    try {
      await this.validateProps()
      const continueTransfer = await swal({
        buttons: ['Cancel', 'Transfer'],
        dangerMode: true,
        icon: 'warning',
        text: `
          Once submitted, the transaction can not be reverted!
          The fee will be ${await this.props.getFee(Number(this.props.amount))} Kinesis
        `,
        title: 'Continue with transfer?',
      })

      if (!continueTransfer) {
        return
      }
      this.props.transferRequest(this.props)
    } catch (e) {
      formAlert(e.message, e.key)
    }
  }

  validateProps = async () => {
    this.checkWalletIsUnlocked()
    this.checkValidTarget()
    await this.checkValidAmount()
    this.checkValidMemo()
  }

  checkWalletIsUnlocked = () => {
    if (!this.props.isWalletUnlocked) {
      throw new WalletLockError()
    }
  }

  checkValidTarget = () => {
    if (!this.props.targetPayee) {
      throw new InputError(`Payee is required`, `transfer-target-address`)
    }
  }

  checkValidAmount = async () => {
    const fee = await this.props.getFee(Number(this.props.amount))

    if (Number(fee) + Number(this.props.amount) > Number(this.props.accountBalance)) {
      throw new InputError(
        `Transfer amount (including ${fee} fee) is higher than your account balance`,
        'transfer-amount',
      )
    } else if (Number(this.props.amount) <= 0) {
      throw new InputError('Transfer amount must be greater than 0', 'transfer-amount')
    } else if (!this.props.amount) {
      throw new InputError('Amount is required', 'transfer-amount')
    }
  }

  checkValidMemo = () => {
    if (this.props.memo.length > 24) {
      throw new InputError('Memo must be fewer than 25 characters', 'transfer-memo')
    }
  }

  isPayeeSelected = () => {
    return (
      this.props.payees.findIndex(({ publicKey }) => publicKey === this.props.targetPayee) !== -1
    )
  }

  copyTransferTransaction = async () => {
    const transaction = await generateTransferTransaction(
      this.props.publicKey,
      this.props.connection,
      this.props,
    )
    copy(
      transaction
        .toEnvelope()
        .toXDR()
        .toString('base64'),
    )
  }

  render() {
    const { updateTransferForm: handleChange } = this.props
    return (
      <div style={{ position: 'relative' }}>
        {this.props.isTransferring && <Loader />}
        <div style={this.props.isTransferring ? { filter: 'blur(2px)' } : {}}>
          <PayeeSelector
            handleChange={handleChange}
            payees={this.props.payees}
            targetPayee={this.props.targetPayee}
            changeTransferView={this.props.changeTransferView}
          />
          <InputField
            value={this.props.targetPayee}
            id="transfer-target-address"
            icon="fa-address-card"
            placeholder="Or input target address"
            disabled={this.isPayeeSelected()}
            onChangeHandler={newValue => handleChange({ field: 'targetPayee', newValue })}
          />
          <InputField
            value={this.props.amount}
            id="transfer-amount"
            icon="fa-coins"
            type="number"
            placeholder="Amount"
            onChangeHandler={newValue => handleChange({ field: 'amount', newValue })}
          />
          <InputField
            value={this.props.memo}
            id="transfer-memo"
            icon="fa-comment"
            placeholder="Optional message"
            onChangeHandler={newValue => handleChange({ field: 'memo', newValue })}
          />
          <div className="field is-grouped">
            <div className="control is-expanded">
              <button className="button is-fullwidth" onClick={this.initTransfer}>
                <span>Transfer</span>
              </button>
            </div>
            <div className="control">
              <button
                className="button is-text"
                onClick={this.copyTransferTransaction}
                title="Copy Transaction"
              >
                <span className="icon">
                  <i className="fas fa-copy" />
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    )
  }
}
