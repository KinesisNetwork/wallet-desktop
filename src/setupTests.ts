import { configure } from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'

export default configure({ adapter: new Adapter() })
