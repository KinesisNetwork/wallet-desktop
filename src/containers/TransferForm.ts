import { connect } from 'react-redux'

import { changeTransferView, changeWalletView, transferRequest, updateTransferForm } from '@actions'
import { TransferForm as TransferPresentation } from '@components/TransferForm'
import { getActiveAccount, getCurrentConnection } from '@selectors'
import { getFeeInKinesis } from '@services/kinesis'
import { RootState } from '@store'
import { Contact } from '@types'

export const mapStateToProps = (state: RootState) => {
  const { wallets, connections, transfer, accounts, contacts } = state
  const { keypair } = getActiveAccount(state.wallet)
  const { privateKey, publicKey } = {
    publicKey: keypair.publicKey(),
    privateKey: keypair.secret(),
  }
  const otherWalletsAsPayees: Contact[] = wallets.walletList
    .filter(wallet => wallet.publicKey !== publicKey)
    .map((wallet): Contact => ({ name: wallet.accountName, address: wallet.publicKey }))
  return {
    ...transfer.form,
    getFee: (amount: number) => getFeeInKinesis(getCurrentConnection(connections), amount),
    isTransferring: transfer.isTransferring,
    isWalletUnlocked: !!privateKey,
    accountBalance: accounts.accountInfo.balance,
    payees: contacts.contactList.concat(otherWalletsAsPayees),
    publicKey,
    connection: getCurrentConnection(connections),
  }
}

const mapDispatchToProps = {
  changeWalletView,
  changeTransferView,
  transferRequest,
  updateTransferForm,
}

export type TransferProps = ReturnType<typeof mapStateToProps> & typeof mapDispatchToProps
export const TransferForm = connect(
  mapStateToProps,
  mapDispatchToProps,
)(TransferPresentation)
