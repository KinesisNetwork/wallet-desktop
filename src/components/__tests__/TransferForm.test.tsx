import { shallow } from 'enzyme';
import * as React from 'react'

import { TransferForm } from '@components/TransferForm'
import '../../setupTests'

describe('TransferForm', () => {
  let props

  beforeEach(() => {
    props = {
      accountBalance: '2',
      activeWallet: {},
      amount: '0.1',
      changeTransferView: () => null,
      changeWalletView: () => null,
      connection: {},
      getFee: () => null,
      invalidForm: () => null,
      isTransferring: false,
      isWalletUnlocked: true,
      memo: '',
      payees: [
        {
          name: 'Fred',
          publicKey: '123'
        }
      ],
      targetPayee: 'asdf',
      transferRequest: () => null,
      updateTransferForm: () => null
    }
  })

  it('renders correctly', () => {
    const wrapper = shallow(<TransferForm {...props} />)

    expect(wrapper).toMatchSnapshot()
  })

  it('displays Loader component if isTransferring is true', () => {
    const wrapper = shallow(<TransferForm {...props} isTransferring={true} />)

    expect(wrapper.find('Loader')).toHaveLength(1)
  })

  it('renders a PayeeSelector, three InputField components and two buttons', () => {
    const wrapper = shallow(<TransferForm {...props} />)

    const button = wrapper.find('button')

    expect(wrapper.find('PayeeSelector')).toHaveLength(1)
    expect(wrapper.find('InputField')).toHaveLength(3)
    expect(button).toHaveLength(2)
    expect(button.at(0).text()).toBe('Transfer')
    expect(button.at(1).find('i').hasClass('fa-copy')).toBe(true)
    expect(button.at(1).prop('title')).toEqual('Copy Transaction')
  })
})
