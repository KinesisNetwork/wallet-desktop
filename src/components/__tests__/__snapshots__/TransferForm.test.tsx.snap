// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransferForm renders correctly 1`] = `
<div
  style={
    Object {
      "position": "relative",
    }
  }
>
  <div
    style={Object {}}
  >
    <PayeeSelector
      changeTransferView={[Function]}
      handleChange={[Function]}
      payees={
        Array [
          Object {
            "name": "Fred",
            "publicKey": "123",
          },
        ]
      }
      targetPayee="asdf"
    />
    <InputField
      icon="fa-address-card"
      id="transfer-target-address"
      isDisabled={false}
      onChangeHandler={[Function]}
      placeholder="Or input target address"
      value="asdf"
    />
    <InputField
      icon="fa-coins"
      id="transfer-amount"
      onChangeHandler={[Function]}
      placeholder="Amount"
      type="number"
      value="0.1"
    />
    <InputField
      icon="fa-comment"
      id="transfer-memo"
      onChangeHandler={[Function]}
      placeholder="Optional message"
      value=""
    />
    <div
      className="field is-grouped"
    >
      <div
        className="control is-expanded"
      >
        <button
          className="button is-fullwidth"
          onClick={[Function]}
        >
          <span>
            Transfer
          </span>
        </button>
      </div>
      <div
        className="control"
      >
        <button
          className="button is-text"
          onClick={[Function]}
          title="Copy Transaction"
        >
          <span
            className="icon"
          >
            <i
              className="fas fa-copy"
            />
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
